UICatalog/Scenarios/VirtualContent.cs

```diff
@@ -0,0 +1,343 @@
+            {
+                view.ViewportSettings &= ~ViewportSettings.ClearVisibleContentOnly;
+            }
+        }
+
+        view.Padding.Add (labelContentSize, contentSizeWidth, labelComma, contentSizeHeight, cbClearOnlyVisible);
+
+        // Add demo views to show that things work correctly
+        var textField = new TextField { X = 20, Y = 7, Width = 15, Text = "Test TextField" };
+
+        var colorPicker = new ColorPicker { Title = "BG", BoxHeight = 1, BoxWidth = 1, X = Pos.AnchorEnd (11), Y = 10 };
+        colorPicker.BorderStyle = LineStyle.RoundedDotted;
+
+        colorPicker.ColorChanged += (s, e) =>
+                                    {
+                                        colorPicker.SuperView.ColorScheme = new (colorPicker.SuperView.ColorScheme)
+                                        {
+                                            Normal = new (
+                                                          colorPicker.SuperView.ColorScheme.Normal.Foreground,
+                                                          e.Color
+                                                         )
+                                        };
+                                    };
+
+        var textView = new TextView
+        {
+            X = Pos.Center (),
+            Y = 10,
+            Title = "TextView",
+            Text = "I have a 3 row top border.\nMy border inherits from the SuperView.\nI have 3 lines of text with room for 2.",
+            AllowsTab = false,
+            Width = 30,
+            Height = 6 // TODO: Use Dim.Auto
+        };
+        textView.Border.Thickness = new (1, 3, 1, 1);
+
+        var charMap = new CharMap
+        {
+            X = Pos.Center (),
+            Y = Pos.Bottom (textView) + 1,
+            Width = 30,
+            Height = 10
+        };
+
+        charMap.Accept += (s, e) =>
+                              MessageBox.Query (20, 7, "Hi", $"Am I a {view.GetType ().Name}?", "Yes", "No");
+
+        var buttonAnchoredRight = new Button { X = Pos.AnchorEnd (10), Y = 0, Text = "Button" };
+
+        var labelAnchoredBottomLeft = new Label
+        {
+            AutoSize = false,
+            Y = Pos.AnchorEnd (3),
+            Width = 25,
+            Height = Dim.Fill (),
+            Text = "Label\nY=AnchorEnd(3),Height=Dim.Fill()"
+        };
+
+        view.Margin.Data = "Margin";
+        view.Margin.Thickness = new (0);
+
+        view.Border.Data = "Border";
+        view.Border.Thickness = new (3);
+
+        view.Padding.Data = "Padding";
+
+        view.Add (buttonAnchoredRight, textField, colorPicker, charMap, textView, labelAnchoredBottomLeft);
+
+        var longLabel = new Label
+        {
+            Id = "label2",
+            X = 0,
+            Y = 30,
+            Text =
+                "This label is long. It should clip to the Viewport (but not ContentArea). This is a virtual scrolling demo. Use the arrow keys and/or mouse wheel to scroll the content."
```

Comment:
Here's another example that shouldn't happen. That's why I insisted that `LayoutViews` and `SetRelativeLayout` should ensure that this doesn't happen.
Another solution is to prevent it from being drawn outside of `ContentSize` but it could be more complicated, I think.

![image](https://github.com/gui-cs/Terminal.Gui/assets/13117724/ad04ed1e-acdb-48a5-a612-ac9fc9c3ceaf)

Reply:
This is a great topic! I thought about this, then lost my train of thought and then moved on leaving the text in that label there (i think in the hopes it would remind me to come back to it; LOL). Let's figure it out...

I'm now thinking more and running some experiments. I'll come back on this once I get my head straight,. 

Thanks for pointing it out.

Reply:
> Here's another example that shouldn't happen. That's why I insisted that `LayoutViews` and `SetRelativeLayout` should ensure that this doesn't happen. Another solution is to prevent it from being drawn outside of `ContentSize` but it could be more complicated, I think.

![HUeeYrN 1](https://github.com/gui-cs/Terminal.Gui/assets/585482/4681538a-fae4-420a-ae1b-26bcc10ed8a5)

Reply:
I do see the cursor bug. Looking at that now.

Reply:
> I do see the cursor bug. Looking at that now.

It's related to not yet supporting `CanFocus = true` in `Adornment`s. I'm going to punt fixing it for now.

Reply:
> Does this work?

What's the definition for `ViewportSettings.DoNotClipContent` flag?

Reply:
In any case, when resetting the `ContentSize` you must re-layout the views to accommodate the new size. It's not just preventing drawing or clipping content outside the `ContentSize` zone, but recalculating the locations and dimensions of the views. Only what I just mentioned is that it is valid for content between the location and the size of the `ContentArea`.

Reply:
Otherwise, how can you simulate a zoom without this effect? Just by clipping you are not readjusting and resizing the views.

Reply:
The content that does not belong to the `ContentArea` is just the background of the visualization, which can be empty, or if the user wants, they can draw a logo or whatever they want. What I just mentioned is suddenly an idea in which the `View.Clear` method will be virtual for the user to draw whatever they want. So, if the `ContentArea` is smaller or offset, it is this background that will be seen, drawn by `Clear`.

Reply:
Diving into this, and writing some new primitive Clip unit tests I have been reminded that clipping is/has been fundamentally broken and works today only mostly/by accident.

This is going to require some deep work to get right. I am now going to focus on this problem as it is critical!

Really glad you pointed this problem out!

Reply:
> Diving into this, and writing some new primitive Clip unit tests I have been reminded that clipping is/has been fundamentally broken and works today only mostly/by accident.
> 
I think it's not just a question of clipping but of layout. You are clipping where you don't want certain content to be drawn, but some of the views have a Computed layout and need to recalculate their location and dimension when resizing the ContentSize and you won't be able to solve this by manipulating the clip but rather through re-layout.

> This is going to require some deep work to get right. I am now going to focus on this problem as it is critical!
> 
Yes, the re-layout problem really needs to be resolved.

> Really glad you pointed this problem out!

It's always a pleasure and I'm always available whenever needed.